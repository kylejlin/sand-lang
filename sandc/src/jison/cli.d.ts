export function main(opts: JisonCliOptions): void;

export interface JisonCliOptions {
  file: string;
  outfile?: string;
}
