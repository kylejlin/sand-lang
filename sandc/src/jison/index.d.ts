export class Parser {
  constructor(grammar: any);

  lexer: Scanner;
  yy: any;

  generate(): string;
  parse(program: string): any;
}

export function print(line: string): void;

export interface Scanner<TokenType = string> {
  yytext: string;
  yyleng: number;
  yylloc: JisonNodeLocation;
  /**
   * For whatever odd reason, Jison starts yylineno at 0,
   * so make sure to subtract one if your line count is 1-based.
   */
  yylineno: number;

  lex(): TokenType;
  setInput(input: string): void;

  pastInput(): string;
  upcomingInput(): string;
  input(): void;
}

export interface JisonNodeLocation {
  first_line: number;
  last_line: number;
  first_column: number;
  last_column: number;
}
