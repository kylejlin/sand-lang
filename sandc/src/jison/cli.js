const jisonCli = require("jison/lib/cli");

module.exports = jisonCli;
