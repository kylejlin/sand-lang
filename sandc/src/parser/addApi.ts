import {
  BinaryExpr,
  BinaryOperation,
  camelCase,
  ConstraintType,
  IfAlternativeType,
  JisonNodeLocation,
  merge,
  NodeType,
  UnaryExpr,
  UnaryOperation,
  Expr,
  DotExpr,
} from "../ast";
import { wrapPrimitiveIfNeeded } from "../sandTypes";
import { SandParser } from "./parser.generated";

export default function addApi(parser: SandParser) {
  const { yy } = parser;

  yy.NodeType = NodeType;
  yy.IfAlternativeType = IfAlternativeType;
  yy.ConstraintType = ConstraintType;

  yy.binaryExpr = function binaryExpr(
    operation: BinaryOperation,
    left: any,
    right: any,
    location: JisonNodeLocation,
  ): BinaryExpr {
    return {
      type: NodeType.BinaryExpr,
      operation,
      left,
      right,
      location: camelCase(location),
    };
  };

  yy.unaryExpr = function unaryExpr(
    operation: UnaryOperation,
    right: any,
    location: JisonNodeLocation,
  ): UnaryExpr {
    return {
      type: NodeType.UnaryExpr,
      operation,
      right,
      location: camelCase(location),
    };
  };

  yy.dotExpr = function dotExpr(
    left: Expr,
    right: string,
    location: JisonNodeLocation,
  ): DotExpr {
    return {
      type: NodeType.DotExpr,
      left,
      right,
      location: camelCase(location),
    };
  };

  yy.wrapPrimitiveIfNeeded = wrapPrimitiveIfNeeded;
  yy.camelCase = camelCase;
  yy.merge = merge;
}
