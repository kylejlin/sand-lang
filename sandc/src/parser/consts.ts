import path from "path";

export const GRAMMAR_PATH = path.join(__dirname, "../../grammar/sand.jison");
