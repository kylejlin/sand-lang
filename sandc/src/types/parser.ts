import { Option, Result } from "rusty-ts";
import { TextRange } from ".";
import * as ast from "./ast";
import { TokenType } from "./tokens";

export interface SandParser {
  parse(input: string): Result<ast.FileNode, ParseError>;
}

export type ParseError = UnexpectedTokenError;

export enum ParseErrorType {
  UnexpectedToken,
  DuplicatePropertyAccessorDeclarations,
}

export interface UnexpectedTokenError {
  errorType: ParseErrorType.UnexpectedToken;
  token: TokenType;
  text: Option<string>;
  location: TextRange;
  expected: TokenType[];
}
