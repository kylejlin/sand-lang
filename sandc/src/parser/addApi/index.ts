import {
  camelCase,
  ConstraintType,
  DotExpr,
  Expr,
  Identifier,
  IfAlternativeType,
  InfixExpr,
  InfixOperation,
  JisonNodeLocation,
  merge,
  NodeType,
  PrefixExpr,
  PrefixOperation,
  Type,
  NodeLocation,
} from "../../ast";
import { wrapPrimitiveIfNeeded } from "../../sandTypes";
import { SandParser } from "../parser.generated";
import typeParser from "../subparsers/type/prebuilt";
import getUpcomingObjectLiteralType from "./lexUtils/getUpcomingObjectLiteralType";
import isThereUpcomingTypeArgListAndOpenParen from "./lexUtils/isThereUpcomingTypeArgListAndOpenParen";

interface PointLocation {
  line: number;
  column: number;
}

export default function addApi(parser: SandParser) {
  const { yy } = parser;

  yy.lexUtils = {};

  yy.lexUtils.isThereUpcomingTypeArgListAndOpenParen = isThereUpcomingTypeArgListAndOpenParen;
  yy.lexUtils.getUpcomingObjectLiteralType = getUpcomingObjectLiteralType;

  yy.NodeType = NodeType;
  yy.IfAlternativeType = IfAlternativeType;
  yy.ConstraintType = ConstraintType;

  yy.binaryExpr = function binaryExpr(
    operation: InfixOperation,
    left: any,
    right: any,
    location: JisonNodeLocation,
  ): InfixExpr {
    return {
      type: NodeType.InfixExpr,
      operation,
      left,
      right,
      location: camelCase(location),
    };
  };

  yy.unaryExpr = function unaryExpr(
    operation: PrefixOperation,
    right: any,
    location: JisonNodeLocation,
  ): PrefixExpr {
    return {
      type: NodeType.PrefixExpr,
      operation,
      right,
      location: camelCase(location),
    };
  };

  yy.dotExpr = function dotExpr(
    left: Expr,
    right: string,
    location: JisonNodeLocation,
  ): DotExpr {
    return {
      type: NodeType.DotExpr,
      left,
      right,
      location: camelCase(location),
    };
  };

  yy.buildDotChainIfNeeded = function buildDotChainIfNeeded(
    idents: Identifier[],
  ): Identifier | DotExpr {
    if (idents.length === 0) {
      throw new Error(
        "Cannot build dot chain or identifier out of empty array.",
      );
    }

    if (idents.length === 1) {
      return idents[0];
    }

    const leftmostDot: DotExpr = {
      type: NodeType.DotExpr,
      left: idents[0],
      right: idents[1].name,
      location: merge(idents[0].location, idents[1].location),
    };

    return idents
      .slice(2)
      .reduce((left: DotExpr, right: Identifier): DotExpr => {
        return {
          type: NodeType.DotExpr,
          left,
          right: right.name,
          location: merge(left.location, right.location),
        };
      }, leftmostDot);
  };

  yy.parseType = function parseType(
    src: string,
    location: JisonNodeLocation,
  ): Type {
    const start = { line: location.first_line, column: location.first_column };
    const incorrectlyLocated = typeParser.parse(src);
    return applyLocationCorrections(incorrectlyLocated, start);
  };

  function applyLocationCorrections(node: Type, start: PointLocation): Type {
    return {
      type: NodeType.Type,
      name: node.name,
      args: node.args.map(subType => applyLocationCorrections(subType, start)),
      location: getAbsoluteNodeLocation(node.location, start),
    };
  }

  /**
   * Takes a location relative to its start position (offset)
   * and computes the absolute location.
   */
  function getAbsoluteNodeLocation(
    relative: NodeLocation,
    offset: PointLocation,
  ): NodeLocation {
    const absStart = getAbsolutePointLocation(getStart(relative), offset);
    const absEnd = getAbsolutePointLocation(getEnd(relative), offset);
    return mergePointLocations(absStart, absEnd);
  }

  function getStart(location: NodeLocation): PointLocation {
    return {
      line: location.firstLine,
      column: location.firstColumn,
    };
  }

  function getEnd(location: NodeLocation): PointLocation {
    return {
      line: location.lastLine,
      column: location.lastColumn,
    };
  }

  function getAbsolutePointLocation(
    relative: PointLocation,
    offset: PointLocation,
  ): PointLocation {
    if (relative.line === 1) {
      return {
        line: relative.line + offset.line - 1,
        column: relative.column + offset.column,
      };
    } else {
      return {
        line: relative.line + offset.line - 1,
        column: relative.column,
      };
    }
  }

  function mergePointLocations(
    start: PointLocation,
    end: PointLocation,
  ): NodeLocation {
    return {
      firstLine: start.line,
      firstColumn: start.column,
      lastLine: end.line,
      lastColumn: end.column,
    };
  }

  yy.wrapPrimitiveIfNeeded = wrapPrimitiveIfNeeded;
  yy.camelCase = camelCase;
  yy.merge = merge;
}
