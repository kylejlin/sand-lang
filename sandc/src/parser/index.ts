import { option, Result, result } from "rusty-ts";
import { JisonUnexpectedTokenError } from "../jison";
import { convertToTextRange } from "../textRange";
import * as ast from "../types/ast";
import { ParseError, ParseErrorType, SandParser } from "../types/parser";
import { TokenType } from "../types/tokens";
import addApi from "./addApi";
import {
  parser as generatedParser,
  SandParserGeneratedByJison,
} from "./parser.generated";

addApi(generatedParser);

const wrappedParser = wrapGeneratedParser(generatedParser);

export default wrappedParser;

function wrapGeneratedParser(
  generated: SandParserGeneratedByJison,
): SandParser {
  return {
    parse(input: string): Result<ast.FileNode, ParseError> {
      try {
        return result.ok(generated.parse(input));
      } catch (e) {
        if (!("object" === typeof e && e !== null)) {
          throw new Error("Unhandled error: " + e);
        }

        if ("object" === typeof e.hash) {
          const {
            hash: { token, text, loc, expected },
          }: JisonUnexpectedTokenError<TokenType> = e;
          return result.err({
            errorType: ParseErrorType.UnexpectedToken,
            token,
            text: text === undefined ? option.none() : option.some(text),
            location: convertToTextRange(loc),
            expected,
          });
        }

        throw new Error("Unhandled error: " + e);
      }
    },
  };
}
