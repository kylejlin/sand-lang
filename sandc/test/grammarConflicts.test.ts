import jison from "../src/jison";
import fs from "fs";
import { GRAMMAR_PATH } from "../src/parser/consts";

describe("sand.jison", () => {
  const jisonLogs = getJisonLogs();
  const conflictCount = getConflictCount(jisonLogs);

  test("sand.jison has the same conflicts", () => {
    expect(jisonLogs).toMatchSnapshot();
  });

  test("sand.jison has the same number of each type of conflict", () => {
    expect(conflictCount).toMatchSnapshot();
  });
});

function getJisonLogs(): string {
  let out = "";
  function print(line: string) {
    out += line;
    out += "\n";
  }

  jison.print = print;

  const src = fs.readFileSync(GRAMMAR_PATH, "utf8");
  const _parser = new jison.Parser(src);

  return out;
}

function getConflictCount(logs: string): ConflictCount {
  const matches = logs.match(
    /Conflict in grammar[\s\S]*?(?=(?:Conflict in grammar|States with conflicts|$))/g,
  );

  if (matches === null) {
    return { total: 0, reduceShift: 0, reduceReduce: 0 };
  } else {
    return {
      total: matches.length,
      reduceShift: matches.filter(isReduceShift).length,
      reduceReduce: matches.filter(isReduceReduce).length,
    };
  }
}

function isReduceShift(match: string): boolean {
  return /- reduce[\s\S]*- shift/.test(match);
}

function isReduceReduce(match: string): boolean {
  return /- reduce[\s\S]*- reduce/.test(match);
}

interface ConflictCount {
  total: number;
  reduceShift: number;
  reduceReduce: number;
}
