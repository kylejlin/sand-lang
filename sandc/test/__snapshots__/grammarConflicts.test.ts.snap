// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sand.jison sand.jison has the same conflicts 1`] = `""`;

exports[`sand.jison sand.jison has the same number of each type of conflict 1`] = `
Object {
  "reduceReduce": 0,
  "reduceShift": 0,
  "total": 0,
}
`;
